File: lib/domain/crop/interactors/masters_nutrient_requirement_create_interactor.rb
# frozen_string_literal: true

module Domain
  module Crop
    module Interactors
      class MastersNutrientRequirementCreateInteractor
        def initialize(output_port:, gateway:, logger:)
          @output_port = output_port
          @gateway = gateway
          @logger = logger
        end

        def call(input_dto)
          unless @gateway.find_nutrient_requirement(input_dto.stage_id).nil?
            @output_port.on_already_exists
            return
          end

          requirement = @gateway.create_nutrient_requirement(input_dto.stage_id, input_dto)
          @output_port.on_create_success(requirement)
        rescue Domain::Shared::Exceptions::RecordInvalid => e
          @output_port.on_validation_errors(e.flatten_error_messages)
        end
      end
    end
  end
end

File: crates/agrr-domain/src/crop/interactors/masters_nutrient_requirement_create_interactor.rs
//! Ruby: `Domain::Crop::Interactors::MastersNutrientRequirementCreateInteractor`
use crate::crop::dtos::NutrientRequirementUpdateInput;
use crate::crop::gateways::{CropGateway, NutrientRequirementGateway};
use crate::crop::ports::MastersNutrientRequirementOutputPort;
use crate::shared::exceptions::RecordInvalidError;

pub struct MastersNutrientRequirementCreateInteractor<'a, G, RG, O> {
    output_port: &'a mut O,
    gateway: &'a G,
    requirement_gateway: &'a RG,
}

impl<'a, G, RG, O> MastersNutrientRequirementCreateInteractor<'a, G, RG, O>
where
    G: CropGateway,
    RG: NutrientRequirementGateway,
    O: MastersNutrientRequirementOutputPort,
{
    pub fn new(output_port: &'a mut O, gateway: &'a G, requirement_gateway: &'a RG) -> Self {
        Self {
            output_port,
            gateway,
            requirement_gateway,
        }
    }

    pub fn call(
        &mut self,
        input: NutrientRequirementUpdateInput,
    ) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        if self
            .requirement_gateway
            .find_by_crop_stage_id(input.stage_id)?
            .is_some()
        {
            self.output_port.on_already_exists();
            return Ok(());
        }
        match self.gateway.create_nutrient_requirement(input.stage_id, input) {
            Ok(entity) => {
                self.output_port.on_create_success(entity);
                Ok(())
            }
            Err(err) => {
                if let Some(e) = err.downcast_ref::<RecordInvalidError>() {
                    self.output_port.on_validation_errors(e.flatten_error_messages());
                    Ok(())
                } else {
                    Err(err)
                }
            }
        }
    }
}
